internal/testutil/acc/atlas.go

```diff
@@ -0,0 +1,112 @@
+package acc
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"runtime"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/atlas-sdk/v20231115007/admin"
+)
+
+// TestMainExecution must be called from TestMain in the test package if ProjectIDExecution is going to be used.
+func TestMainExecution(m *testing.M) {
+	atlasInfo.init = true
+	atlasInfo.resourceName = resourceName()
+
+	exitCode := m.Run()
+
+	if atlasInfo.needsDeletion {
+		fmt.Printf("Deleting execution project: %s, resource: %s\n", atlasInfo.projectName, atlasInfo.resourceName)
+		deleteProject(atlasInfo.projectID)
+	}
+
+	os.Exit(exitCode)
+}
+
+// ProjectIDExecution returns a project id created for the execution of the resource tests.
+func ProjectIDExecution(tb testing.TB) string {
+	tb.Helper()
+	SkipInUnitTest(tb)
+	require.True(tb, atlasInfo.init, "TestMainExecution must called to be able to use ProjectIDExecution")
+
+	atlasInfo.mu.Lock()
+	defer atlasInfo.mu.Unlock()
+
+	// lazy creation so it's only done if really needed
+	if atlasInfo.projectName == "" {
+		var globalName, globalID string
+		if atlasInfo.resourceName != "" {
+			globalName = (prefixProjectKeep + "-" + atlasInfo.resourceName)[:projectNameMaxLen]
+			globalID = projectID(globalName)
+		}
+
+		if globalID == "" {
+			atlasInfo.projectName = RandomProjectName()
+			tb.Logf("Creating execution project: %s, resource: %s, global project (not found): %s\n", atlasInfo.projectName, atlasInfo.resourceName, globalName)
+			atlasInfo.projectID = createProject(tb, atlasInfo.projectName)
+			atlasInfo.needsDeletion = true
+		} else {
+			atlasInfo.projectName = globalName
+			tb.Logf("Reusing global project: %s, resource: %s\n", atlasInfo.projectName, atlasInfo.resourceName)
+			atlasInfo.projectID = globalID
+		}
+	}
+
+	return atlasInfo.projectID
+}
+
+var atlasInfo = struct {
+	projectID     string
+	projectName   string
+	resourceName  string
+	mu            sync.Mutex
+	init          bool
+	needsDeletion bool
+}{}
+
+const (
+	projectNameMaxLen = 64
+)
+
+func resourceName() string {
```

Comment:
I am probably on the side of making things as simple as possible. Can't we just pass the name of the resource when things are initialised?

Reply:
i think this is less error prone than passing the resource name as a string in each call. i think it's better if it can be got automatically.

(something similar happens in the GH Actions where we have some errors from time to time because copying/pasting incorrectly to pass redundant info)

Reply:
ok noted, this method makes me thinking that even if this is code written for testing infra I think it should be unit tested. If we don't test it, I suspect it will become hard to touch for other engineers.

Reply:
also this method is not really very important, it could fail and return "" and everything would work (except trying to reuse the global project for that resource).

but you have a point, let me see if it can be tested, it's a bit tricky because it depends on the resource test structure.